assets/js/components/ProjectTextEditor.js

```diff
@@ -0,0 +1,237 @@
+import $ from 'jquery'
+import { MDCSelect } from '@material/select'
+import { ProgramEditorDialog } from '../custom/ProgramEditorDialog'
+import { CustomTranslationApi } from '../api/CustomTranslationApi'
+
+/* global Routing */
+export function ProjectTextEditor (projectDescriptionCredits, defaultText, programId) {
+  const self = this
+  this.defaultText = defaultText
+  this.programId = programId
+
+  this.editTextUI = $('#edit-text-ui')
+  this.editText = $('#edit-text')
+  this.editTextError = $('#edit-text-error')
+  this.languageSelectorList = $('#edit-language-selector-list')
+  this.selectedLanguage = $('#edit-selected-language')
+  this.textLoadingSpinner = $('#edit-loading-spinner')
+
+  this.languageSelect = new MDCSelect(document.querySelector('#edit-language-selector'))
+  this.languages = {}
+  this.previousIndex = 0
+
+  this.translationSaved = projectDescriptionCredits.data('trans-translation-saved')
+  this.translationDeleted = projectDescriptionCredits.data('trans-translation-deleted')
+
+  this.closeEditorDialog = new ProgramEditorDialog(
+    projectDescriptionCredits.data('trans-close-editor-prompt'),
+    projectDescriptionCredits.data('trans-save'),
+    projectDescriptionCredits.data('trans-discard')
+  )
+
+  this.keepOrDiscardDialog = new ProgramEditorDialog(
+    projectDescriptionCredits.data('trans-save-on-language-change'),
+    projectDescriptionCredits.data('trans-keep'),
+    projectDescriptionCredits.data('trans-discard')
+  )
+
+  $('#edit-submit-button').on('click', () => { this.save() })
+  $('#edit-close-button').on('click', () => { this.close() })
+
+  this.languageSelect.listen('MDCSelect:change', () => {
+    if (!this.editText.is(':visible') || this.languageSelect.selectedIndex === this.previousIndex) {
+      return
+    }
+
+    if (this.areChangesSaved()) {
+      this.previousIndex = this.languageSelect.selectedIndex
+      this.getCustomTranslation()
+    } else {
+      this.keepOrDiscardDialog.show(keepOrDiscardChangesResult)
+    }
+  })
+
+  $(document).ready(getLanguages)
+
+  this.show = (config, initialText, closedCallback) => {
+    self.programSection = config.programSection
+    this.customTranslationApi = new CustomTranslationApi(config.programSection)
+    self.editText.val(initialText)
+    self.lastSavedText = initialText

```

Comment:
This function has a mix of `this.` and `self.`, but should be one or the other for consistency